MINOR: Greedily initialize allTopicStats BrokerTopicMetrics

stanislavkozlovski opened the pull request:
This patch fixes the quota system test whose JMX tool relies on the metrics existing.

stanislavkozlovski committed:
Greedily initialize allTopicStats BrokerTopicMetrics
This patch fixes the quota system test whose JMX tool relies on the metrics existing.

stanislavkozlovski commented:
cc @ijuma @guozhangwang @tuvtran for reviews

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -188,6 +188,24 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
+    totalProduceRequestRate
+    totalFetchRequestRate
+    fetchMessageConversionsRate
+    produceMessageConversionsRate
```
This is brittle. How do we ensure we don't forget to add an item here when we add a new `def`? We could either add a test to help us or change the code to make this kind of thing impossible. One way would be to replace `def` with `val` and have the laziness be optional within a class that wraps the `getAndMaybePut` logic.

stanislavkozlovski committed:
Change lazy metric definition in BrokerTopicMetrics

We use a case class for lazy metrics which conditionally registers the metric greedily if there are no metric tags present.
Also adds a unit test to verify the expected behavior.
This makes the logic more robust to new changes

stanislavkozlovski commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +146,65 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
-    removeMetricHelper(BrokerTopicStats.FetchMessageConversionsPerSec, tags)
-    removeMetricHelper(BrokerTopicStats.ProduceMessageConversionsPerSec, tags)
-  }
+  def close() = metricTypeMap.values.foreach(_.close())
```
I wasn't sure if we should clear the map here or not.
If we do clear it, there's no going back - nothing will repopulate it.
If we don't, we are saving metric in memory for no reason.

I traced the usages of #close() and they seem to be called when we either deallocate the `BrokerTopicMetrics` instance or shut down the server, so I think it's fine to not clear it. Nevertheless, I'm open to suggestions

ijuma replied:
I don't think we should clear it for the reason I mentioned in the other comment. Or you'd have to add some logic to re-add to the underlying map. But the memory usage of the wrapper without the underlying metric is pretty small.

guozhangwang replied:
With leader partition migrating back and forth, would that cumulate a lot of metrics no-longer-needed?

ijuma replied:
Metrics are closed. We just keep the wrapper.

guozhangwang replied:
I misunderstood the original question and thought @stanislavkozlovski was asking about the `foreach(_.close())` call.

Agree that we do not need to clear the map of wrappers.

stanislavkozlovski commented:
@ijuma do you think the latest commit addresses your comment?

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +146,65 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
-  def bytesRejectedRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.BytesRejectedPerSec,
-    newMeter(BrokerTopicStats.BytesRejectedPerSec, "bytes", TimeUnit.SECONDS, tags))
+  // public for testing only
+  val metricTypeMap = mutable.Map[String, LazyMeter](
```
This has to be thread-safe, right? Same for LazyMeter and `removeMetricsHelper`.

stanislavkozlovski replied:
Great catch. I _think_ the newest implementation should be thread-safe, although it is late so I may have an error somewhere

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +146,65 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
     case Some(topic) => Map("topic" -> topic)
   }
 
+  case class LazyMeter(metricType: String,
```
Maybe we should call this `MeterWrapper` or something since it's not necessarily lazy. Nit: both parameters can be on the same line.

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +146,65 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
-    }
-  }
+  def removeMetricHelper(metricType: String, tags: scala.collection.Map[String, String]): Unit =
+    metricTypeMap.remove(metricType).foreach(_.close())
```
If you remove a metric via this method, how are they added back? I think you cannot remove it from `metricTypeMap`, you have to:
1. Call `close()`
2. Change `close()` to set its var to `null` so that if it's invoked again, it's recreated.

ijuma reviewed (commented):
Thanks for the updates. This is along the lines of what I was thinking, but I left some comments.

stanislavkozlovski committed:
Rename LazyMeter, make it thread-safe and better handle removing metrics

Also adds a new putAll method to Pool.scala

stanislavkozlovski committed:
Fix Pool#putAll to be thread-safe

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +147,83 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
   }
 
+  case class MeterWrapper(metricType: String, eventType: String) {
+    var lazyMeter: Meter = _
```
We should make this private and volatile.

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +147,83 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
+    var lazyMeter: Meter = _
+    private val meterLock = new Object
+
+    def meter(): Meter = meterLock synchronized {
```
Once `lazyMeter` is volatile, you can check for `null` before you acquire the lock so reads don't require a lock.

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +147,83 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
-      removeMetric(metricType, tags)
-    }
+    val metric = metricTypeMap.get(metricType)
+    if (metric != null)
```
Why do we check for `null` here, but not in other places where we do `get`?

stanislavkozlovski replied:
I don't have null checks in the other places because it's very straightforward to see that the values are there. Here, I wasn't sure what calls the function so I just left it be.

Now that I think about it, I see that we have null checks in the places we call `removeMetricHelper()`, so we can remove the check. I don't have a strong opinion on this

stanislavkozlovski committed:
Remove trailing comma

stanislavkozlovski commented:
Scala 2.11 build failed because of a trailing comma

ijuma commented on core/src/test/scala/unit/kafka/metrics/MetricsTest.scala:
```diff
@@ -166,9 +180,22 @@ class MetricsTest extends KafkaServerTestHarness with Logging {
+  }
+
+  private def filterByTopicMetricRegex(metrics: Set[String], topic: Option[String]): Set[String] = {
+    val baseRegexPattern = ".*BrokerTopicMetrics.*"
```
You can use `.r` on the string to create a regex. Also, instead of the pattern match, you can say something like:

```scala
baseRegexPattern + topic.map(t => s"($t)" + "$").getOrElse("")
```

ijuma reviewed (commented):
Thanks for the updates, a few more comments.

stanislavkozlovski committed:
Address PR comments

Make the variable volatile and private, and do not acquire lock in the common read case

guozhangwang commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +147,84 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
-      removeMetric(metricType, tags)
-    }
-  }
+  def removeMetricHelper(metricType: String, tags: scala.collection.Map[String, String]): Unit =
```
Do we still need this helper function (the tags are no longer needed)? The only left caller `removeOldLeaderMetrics` do not need the tags.

tuvtran replied:
`removeOldLeaderMetrics` doesn't need the tags but it doesn't delete all the metrics so I think we could change this helper function into a method that removes one type of metric vs `close()` that removes all metrics

stanislavkozlovski replied:
I removed the tags argument and renamed it to `closeMetric`

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +147,84 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
+            lazyMeter = newMeter(metricType, eventType, TimeUnit.SECONDS, tags)
+        }
+      }
+      lazyMeter
```
You have to read `lazyMeter` into a local variable before the null check. Otherwise `null` could be returned if  `close` is called at the right time.

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +147,84 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
-  def bytesRejectedRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.BytesRejectedPerSec,
-    newMeter(BrokerTopicStats.BytesRejectedPerSec, "bytes", TimeUnit.SECONDS, tags))
+  // public for testing only
+  val metricTypeMap = new Pool[String, MeterWrapper]()
```
We don't want to expose the mutable map. If you need this in tests, expose a read-only accessor.

ijuma commented on core/src/test/scala/unit/kafka/metrics/MetricsTest.scala:
```diff
@@ -75,6 +75,20 @@ class MetricsTest extends KafkaServerTestHarness with Logging {
+    createTopic(topic, 2, 1)
+
+    // The broker metrics for all topics should be greedily registered
+    assertTrue("General topic metrics don't exist", topicMetrics(None).nonEmpty)
```
I'd replace this assert with one that checks one of the map entries.

stanislavkozlovski replied:
What is the reasoning?

ijuma replied:
An assert that fails saying that metrics should be non empty but aren't is less useful than an assert that tells you that a particular metric doesn't exist. Generally, strive to avoid `assertTrue` if there is an alternative since tests that fail in this way tend to be the most annoying. The first thing you have to do is change the assert to get more information.

ijuma replied:
Similar for the `size` thing, if you know what the elements should be, it's better to assert that the collections are the same rather than just checking the `size`. These are guidelines and don't always apply, but something to keep in mind.

stanislavkozlovski replied:
Thanks for the suggestion. I agree and I'll keep these guidelines in mind next time. In this particular instance, it seems hard to do the proper assertion as:
1) MeterWrapper encapsulates the yammer `Meter` object, so unless we make the lazyMeter variable public, we can't really tell whether it's initialized or not (calling meter() would initialize it)
2) We can only access the `MetricName` instance from the default registry (topicMetrics() helper) and our internal map hosts the yammer `Meter`, which doesn't expose names. It's hard to reconcile both

ijuma replied:
Makes sense, thanks for looking into it.

rite2nikhil commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +147,84 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
     case Some(topic) => Map("topic" -> topic)
   }
 
+  case class MeterWrapper(metricType: String, eventType: String) {
```
Can this be generalized to allow other Metric types like `Histogram` ?

stanislavkozlovski replied:
Do you think we should? Considering it's just used for Meters right now, I think we don't need to prematurely generalize it.

ijuma replied:
I think we can leave this for when we need it.

stanislavkozlovski committed:
Fix race condition in meter() method

Rename removeMetricHelper to closeMetric and omit useless tags argument
Make metricTypeMap private and expose a separate method for tests

stanislavkozlovski commented:
Addressed most of the comments and rebased with master

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -146,64 +147,86 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
+        meterLock synchronized {
+          if (lazyMeter == null)
+            lazyMeter = newMeter(metricType, eventType, TimeUnit.SECONDS, tags)
+          meter = lazyMeter
```
Nit: the standard practice is to assign the local variable first and then assign the field from the local variable. It's the same code and a tiny bit more efficient.

ijuma commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -247,21 +270,21 @@ class BrokerTopicStats {
     val topicMetrics = topicStats(topic)
     if (topicMetrics != null)
-      topicMetrics.removeMetricHelper(BrokerTopicStats.ReplicationBytesInPerSec, topicMetrics.tags)
+      topicMetrics.closeMetric(BrokerTopicStats.ReplicationBytesInPerSec, topicMetrics.tags)
```
This doesn't compile.

stanislavkozlovski committed:
Fix compilation error

stanislavkozlovski committed:
Check for null when closing metrics

ijuma reviewed (approved):
LGTM

stanislavkozlovski commented:
**JDK 11/2.13** failing test was
```
org.apache.kafka.connect.integration.RebalanceSourceConnectorsIntegrationTest.testStartTwoConnectors
java.lang.AssertionError: Condition not met within timeout 30000. Connector tasks did not start in time.
```

and is a known flake, tracked in https://issues.apache.org/jira/browse/KAFKA-8661

and I am not sure why the **JDK 11/2.12** failed:

```
23:59:12 * What went wrong:
23:59:12 Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)
23:59:12 
```

ijuma commented:
retest this please

ijuma commented:
Jenkins seems to be stuck. The previous builds failed due to known flakes and the build passed locally. Merging to trunk.
